src/extension/neoExpress/neoExpressInstaller.ts

```diff
@@ -0,0 +1,181 @@
+      vscode.window.showInformationMessage(
+        `The ${this.name} ${current.version} on your box is newer version than the required version:${target.version}. Please check if there is a newer extension.`
+      );
+      return { updateResult: UpdateResult.currentPackageIsNewer, package: current };
+    }
+    if (current.version.compare(target.version) < 0 && current.version.isNewMajorOrMinorVersion(target.version)) {
+      this.context.workspaceState.update(this.neverShowAgainKey, false);
+    }
+    let neverShowAgain = this.context.workspaceState.get<boolean>(this.neverShowAgainKey);
+    if (neverShowAgain) {
+      return { updateResult: UpdateResult.declinedToUpdate, package: current };
+    }
+    return await this.handlePatchUpdate(current, target);
+  }
+
+  private async selectUserPreferredLocation(): Promise<PackageLocation | null> {
+    const moreInfoLink = new vscode.MarkdownString(`[More information](https://dotnet.microsoft.com/download)`);
+    moreInfoLink.isTrusted = true;
+    let response = await vscode.window.showInformationMessage(
+      `${this.name} ${this.version} is required. Where would you like to install it? ${moreInfoLink.value}`,
+      "Local",
+      "Global"
+    );
+    let preferredLocation = null;
+    if (response === "Local") {
+      preferredLocation = PackageLocation.local;
+      Log.log(LOG_PREFIX, `User selected local to install ${this.name}`);
+    } else if (response === "Global") {
+      preferredLocation = PackageLocation.global;
+      Log.log(LOG_PREFIX, `User selected global to install ${this.name}`);
+    } else {
+      Log.log(LOG_PREFIX, `User declined to install ${this.name}. No action taken.`);
+    }
+    return preferredLocation;
+  }
+
+  private async handlePatchUpdate(
+    current: DotNetPackage,
+    target: DotNetPackage
+  ): Promise<{ updateResult: UpdateResult; package: DotNetPackage | null }> {
+    let updatedPackage = current;
+    let updateResult = UpdateResult.notUpdated;
+    // check if required is newer than current
+    const newPackageVersion = await target.version.findLatestPatchVersionFromNuget(
+      getIncludePreviewReleases(),
+      getIncludeBuildServerFeed()
+    );
+    if (current.version.compare(newPackageVersion) >= 0) {
+      return { updateResult: UpdateResult.noNewVersionFromNuget, package: updatedPackage };
+    }
+    // handle update to new version
+    const moreInfoLink = new vscode.MarkdownString(`[More information](https://dotnet.microsoft.com/download)`);
+    moreInfoLink.isTrusted = true;
+    const dontAskAgain = "Dont't ask again";
+    const selection = await vscode.window.showInformationMessage(
+      `Version ${newPackageVersion.toString()} is available. Do you want to update ${locationString(
+        current.location
+      )} to the new version? ${moreInfoLink.value}`,
+      "Yes",
+      "No",
+      dontAskAgain
+    );
+    if (selection === "Yes") {
+      try {
+        await updateCommand(
```

Comment:
ditto to use `window.withProgress` during update

Reply:
added